### ISSUE TITLE
Wrong calculation of Closed findings

### ISSUE THREAD

**Issue by optimistic5** [DefectDojo/django-DefectDojo#11129]
**Bug description**
After reimport DefectDojo closed **all** findings, but not really mitigated.
I had 23 findings (see screenshot 1), then 17 issues was solved and new report was reimported, and I see that calculation is wrong (not expected).
Calculation of **Closed findings** and **Total findings** is wrong.
Two issues:
1. I see 23 **closed findings**, but in reality only 17 was solved. See screenshot 2.
2. I see 29 **total findings**, but I expect 23, because 17 closed + 6 open. See screenshot 2.

**Steps to reproduce**
1. Upload first report with 23 findings.
2. Upload second report with 6 findings (this 6 is the same as in first report)
3. See that closed 23 findings, but not 17.
4. See that Total findings is 29, but not 23 (17 closed and 6 open)

**Expected behavior**
If the finding is the same (same CVE) **it should not be closed** after reimport report.
I expect in **Closed findings** to see only unique closed CVE.

**Deployment method**
- [x] Docker Compose

**Environment information**
 - DefectDojo version **v. 2.39.0**

**Sample scan files**
`report1.json` - has 23 Critical findings
`report2.json` - has 6 Critical findings (6 same as first report has)
`Total findings - DD report.pdf` - Report from DefectDojo where you can see that open CVE is the same as closed.

**Screenshots**
![screen1](https://github.com/user-attachments/assets/35d378a9-34ee-467d-932d-069e85d0cb1d)
![screen2](https://github.com/user-attachments/assets/3d5384ef-d53c-4f3c-b4f8-8ad5354f9fed)

**Additional info**
I use `close_old_findings` argument while reimport. See all arguments I use below.

<details>
  <summary>Script which I use for reimport</summary>
  
```bash
export DOJO_HOST=xyz.com
export DOJO_ENV_NAME=Production
export DOJO_LEAD_ID=1
export DOJO_SCANNER="Trivy Scan"
export DOJO_API_TOKEN=xxx
export DOJO_ENGAGEMENT_NAME="Weekly docker image scanning"
```
run this to import `report1.json` to `aa-ruby-test` project 
```bash
python reimport.py --host ${DOJO_HOST} --api_key ${DOJO_API_TOKEN} --result_file report1.json --scanner "${DOJO_SCANNER}" --engagement_name "${DOJO_ENGAGEMENT_NAME}" --product_name aa-ruby-test --lead_id ${DOJO_LEAD_ID} --environment ${DOJO_ENV_NAME}
```
  
  ```python
  #!/usr/bin/python3

import argparse
from datetime import datetime
import requests
import urllib3
urllib3.disable_warnings()

def upload_results(host, api_key, scanner, result_file, engagement_name, lead_id, environment, verify=False): # set verify to False if ssl cert is self-signed
    API_URL = "https://"+host+"/api/v2"
    IMPORT_SCAN_URL = API_URL+"/reimport-scan/"
    AUTH_TOKEN = "Token "+api_key

    headers = dict()
    json = dict()
    files = dict()

    headers['Authorization'] = AUTH_TOKEN
    print(headers)

    json['minimum_severity'] = "Low"
    json['scan_date'] = datetime.now().strftime("%Y-%m-%d")
    json['verified'] = True
    json['active'] = True
    json['lead'] = lead_id
    json['scan_type'] = scanner
    json['environment'] = environment
    json['close_old_findings'] = True
    json['deduplication_on_engagement'] = True
    json['product_name'] = product_name
    json['product_type_name'] = "microservices"
    json['engagement_name'] = engagement_name
    json['auto_create_context'] = True
    print(json)

    # Prepare file data to send to API
    files['file'] = open(result_file)

    # Make a request to API
    response = requests.post(IMPORT_SCAN_URL, headers=headers, files=files, data=json, verify=verify)
    # Uncomment these lines to debug
    # print(response.request.body)
    # print(response.request.headers)
    # print(response.status_code)
    # print(response.text)
    return response.status_code

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='CI/CD integration for DefectDojo')
    parser.add_argument('--host', help="DefectDojo Hostname", required=True)
    parser.add_argument('--api_key', help="API v2 Key", required=True)
    parser.add_argument('--engagement_name', help="Engagement name", required=True)
    parser.add_argument('--result_file', help="Scanner file", required=True)
    parser.add_argument('--scanner', help="Type of scanner", required=True)
    parser.add_argument('--product_name', help="DefectDojo Product name", required=True)
    parser.add_argument('--lead_id', help="ID of the user conducting the testing", required=True)
    parser.add_argument('--environment', help="Environment name", required=True)
    parser.add_argument('--build_id', help="Reference to external build id", required=False)

    # Parse out arguments
    args = vars(parser.parse_args())
    host = args["host"]
    api_key = args["api_key"]
    product_name = args["product_name"]
    result_file = args["result_file"]
    scanner = args["scanner"]
    engagement_name = args["engagement_name"]
    lead_id = args["lead_id"]
    environment = args["environment"]
    build_id = args["build_id"]

    result = upload_results(host, api_key, scanner, result_file, engagement_name, lead_id, environment)
    if result == 201 :
         print("✅ Successfully uploaded the results to Defect Dojo")
    else:
         print("Something went wrong, please debug " + str(result))
  ```
  
</details>

**Comment by mtesauro** [comment]
@optimistic5 

> If the finding is the same (same CVE) it should not be closed after reimport report.
> I expect in Closed findings to see only unique closed CVE.

DefectDojo doesn't necessarily use only CVE to determine which findings are closed on reimport. What is used is determined on a tool by tool basis.

For Trivy Scans, the fields used to determine the 'same' finding are:
https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/settings/settings.dist.py#L1236

You can adjust how this works by overriding the defaults in a local_settings.py file you create so you can modify it to only use CVE if you want. BTW, CVE would be the first item stored under `"vulnerability_ids"` if a tool provides a CVE. You can find out more about the deduplication algorithms at https://documentation.defectdojo.com/usage/features/#deduplication-algorithms

However, I don't think this is actually a bug if your expectation was that CVE alone was used to close reimported findings.

If you want to more interactively discuss how reimport & dedup works, I'd suggest joining the OWASP Slack and ask in the #defectdojo channel. Details to join are at https://github.com/DefectDojo/django-DefectDojo/tree/master?tab=readme-ov-file#community-getting-involved-and-updates

HTH

**Comment by optimistic5** [comment]
I did investigation.
Here is my findings.
In trivy report of ruby we can see this vulnerability:
<details>
  <summary>vulnerability from ruby 3.1.4</summary>

  
  ```json
  {
          "VulnerabilityID": "CVE-2023-6879",
          "PkgID": "libaom3@3.6.0-1",
          "PkgName": "libaom3",
          "PkgIdentifier": {
            "PURL": "pkg:deb/debian/libaom3@3.6.0-1?arch=amd64\u0026distro=debian-12.5",
            "UID": "b2afd5c5ffd23406"
          },
          "InstalledVersion": "3.6.0-1",
          "Status": "affected",
          "Layer": {
            "Digest": "sha256:b4580645a8e50b87a19330da289a9b1540022379f2c99d3f0112e3c5c4a8d051",
            "DiffID": "sha256:893507f6057f8f0532ae8907fb5470f4a0181181bdbe139c2706d47cc4a88600"
          },
          "SeveritySource": "nvd",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-6879",
          "DataSource": {
            "ID": "debian",
            "Name": "Debian Security Tracker",
            "URL": "https://salsa.debian.org/security-tracker-team/security-tracker"
          },
          "Title": "aom: heap-buffer-overflow on frame size change",
          "Description": "Increasing the resolution of video frames, while performing a multi-threaded encode, can result in a heap overflow in av1_loop_restoration_dealloc().\n\n",
          "Severity": "CRITICAL",
          "CweIDs": [
            "CWE-787",
            "CWE-20"
          ],
          "VendorSeverity": {
            "nvd": 4,
            "redhat": 3
          },
          "CVSS": {
            "nvd": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
              "V3Score": 9.8
            },
            "redhat": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
              "V3Score": 9.8
            }
          },
          "References": [
            "https://access.redhat.com/security/cve/CVE-2023-6879",
            "https://aomedia.googlesource.com/aom/+/refs/tags/v3.7.1",
            "https://crbug.com/aomedia/3491",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AYONA2XSNFMXLAW4IHLFI5UVV3QRNG5K/",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/D6C2HN4T2S6GYNTAUXLH45LQZHK7QPHP/",
            "https://nvd.nist.gov/vuln/detail/CVE-2023-6879",
            "https://www.cve.org/CVERecord?id=CVE-2023-6879"
          ],
          "PublishedDate": "2023-12-27T23:15:07.53Z",
          "LastModifiedDate": "2024-02-02T02:27:15.863Z"
        },
  ```
  
</details>
<details>
  <summary>vulnerability from ruby 3.1.4</summary>

  
  ```json
 {
          "VulnerabilityID": "CVE-2023-6879",
          "PkgID": "libaom3@3.6.0-1+deb12u1",
          "PkgName": "libaom3",
          "PkgIdentifier": {
            "PURL": "pkg:deb/debian/libaom3@3.6.0-1%2Bdeb12u1?arch=amd64\u0026distro=debian-12.7",
            "UID": "ac5b595bbc65735b"
          },
          "InstalledVersion": "3.6.0-1+deb12u1",
          "Status": "affected",
          "Layer": {
            "Digest": "sha256:ad1c7cfc347f5c86fc2678b58f6a8fb6c6003471405760532fc3240b9eb1b343",
            "DiffID": "sha256:d23b5e6144a796978b150684f10a986aa5dc30d4729f5af29864847008211613"
          },
          "SeveritySource": "nvd",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-6879",
          "DataSource": {
            "ID": "debian",
            "Name": "Debian Security Tracker",
            "URL": "https://salsa.debian.org/security-tracker-team/security-tracker"
          },
          "Title": "aom: heap-buffer-overflow on frame size change",
          "Description": "Increasing the resolution of video frames, while performing a multi-threaded encode, can result in a heap overflow in av1_loop_restoration_dealloc().\n\n",
          "Severity": "CRITICAL",
          "CweIDs": [
            "CWE-787",
            "CWE-20"
          ],
          "VendorSeverity": {
            "nvd": 4,
            "redhat": 3
          },
          "CVSS": {
            "nvd": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
              "V3Score": 9.8
            },
            "redhat": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
              "V3Score": 9.8
            }
          },
          "References": [
            "https://access.redhat.com/security/cve/CVE-2023-6879",
            "https://aomedia.googlesource.com/aom/+/refs/tags/v3.7.1",
            "https://crbug.com/aomedia/3491",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AYONA2XSNFMXLAW4IHLFI5UVV3QRNG5K/",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/D6C2HN4T2S6GYNTAUXLH45LQZHK7QPHP/",
            "https://nvd.nist.gov/vuln/detail/CVE-2023-6879",
            "https://www.cve.org/CVERecord?id=CVE-2023-6879"
          ],
          "PublishedDate": "2023-12-27T23:15:07.53Z",
          "LastModifiedDate": "2024-02-02T02:27:15.863Z"
        },
        {
          "VulnerabilityID": "CVE-2023-5841",
          "PkgID": "libopenexr-3-1-30@3.1.5-5",
          "PkgName": "libopenexr-3-1-30",
          "PkgIdentifier": {
            "PURL": "pkg:deb/debian/libopenexr-3-1-30@3.1.5-5?arch=amd64\u0026distro=debian-12.7",
            "UID": "5c104a106ff50b8d"
          },
          "InstalledVersion": "3.1.5-5",
          "Status": "affected",
          "Layer": {
            "Digest": "sha256:ad1c7cfc347f5c86fc2678b58f6a8fb6c6003471405760532fc3240b9eb1b343",
            "DiffID": "sha256:d23b5e6144a796978b150684f10a986aa5dc30d4729f5af29864847008211613"
          },
          "SeveritySource": "nvd",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-5841",
          "DataSource": {
            "ID": "debian",
            "Name": "Debian Security Tracker",
            "URL": "https://salsa.debian.org/security-tracker-team/security-tracker"
          },
          "Title": "OpenEXR: Heap Overflow in Scanline Deep Data Parsing",
          "Description": "Due to a failure in validating the number of scanline samples of a OpenEXR file containing deep scanline data, Academy Software Foundation OpenEX image parsing library version 3.2.1 and prior is susceptible to a heap-based buffer overflow vulnerability. This issue was resolved as of versions v3.2.2 and v3.1.12 of the affected library.\n",
          "Severity": "CRITICAL",
          "CweIDs": [
            "CWE-787",
            "CWE-122"
          ],
          "VendorSeverity": {
            "amazon": 2,
            "nvd": 4,
            "redhat": 3
          },
          "CVSS": {
            "nvd": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:N",
              "V3Score": 9.1
            },
            "redhat": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:N",
              "V3Score": 9.1
            }
          },
          "References": [
            "https://access.redhat.com/security/cve/CVE-2023-5841",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/LSB6DB5LAKGPLRXEF5HDNGUMT7GIFT2C/",
            "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/XWMINVKQLSUHECXBSQMZFCSDRIHFOJJI/",
            "https://nvd.nist.gov/vuln/detail/CVE-2023-5841",
            "https://takeonme.org/cves/CVE-2023-5841.html",
            "https://takeonme.org/cves/CVE-2023-5841.html#",
            "https://www.cve.org/CVERecord?id=CVE-2023-5841"
          ],
          "PublishedDate": "2024-02-01T19:15:08.097Z",
          "LastModifiedDate": "2024-02-26T16:27:49.42Z"
        },
  ```
  
</details>

note this part in ruby:3.1.4:
```json
"VulnerabilityID": "CVE-2023-6879",
"PkgID": "libaom3@3.6.0-1",
"PkgName": "libaom3",
"InstalledVersion": "3.6.0-1",
```

and same part in ruby:3.1.6:
```json
"VulnerabilityID": "CVE-2023-6879",
"PkgID": "libaom3@3.6.0-1+deb12u1",
"PkgName": "libaom3",
"InstalledVersion": "3.6.0-1+deb12u1",
```

`PkgID` and `InstalledVersion` is different. But as for me it's still same finding. Suffix `+deb12u1` does not really change the vulnerability.
But [Trivy Scan has configuration](https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/settings/settings.dist.py#L1237):
`"Trivy Scan": ["title", "severity", "vulnerability_ids", "cwe", "description"],`
where `title` [is](https://github.com/DefectDojo/django-DefectDojo/blob/090b2c7a77a67f02064b158559e5fcdc665c4291/dojo/tools/trivy/parser.py#L202) 
```python
title = f"{vuln_id} {package_name} {package_version}"
```
which in JSON report would be: `VulnerabilityID`, `PkgName`, `InstalledVersion`.

It means that `title` would be different in `ruby:3.1.4` and `ruby:3.1.6`, which is as for me should not be like this.

This is the main thing which was confusing - `title` consist from these parts.

For me I found configuration which suites my needs. I added to `uwsgi` container variables:
```bash
DD_DEDUPLICATION_ALGORITHM_PER_PARSER: '{"Trivy Scan": "hash_code"}'
DD_HASHCODE_FIELDS_PER_SCANNER: '{"Trivy Scan": ["vulnerability_ids", "component_name"]}'
```
And I received desired result:
![image](https://github.com/user-attachments/assets/c4ff5ad6-f879-4d07-bddd-56a919f22a86)

Thank you.

**Comment by mtesauro** [comment]
@optimistic5 Glad you were able to configure DD to fit your needs. Just FYI: You can also keep those dedup changes in local_settings.py but not settings.dist.py - details [here](https://documentation.defectdojo.com/getting_started/configuration/).

Each AppSec/Product Security/DevSecOps place I've worked at or talked to people who worked at different places, the more I've realized that what works at one place won't necessarily work at the next place. Different programming languages, programming styles, environments, ... can impact how you need to use tool output.

We've tried to keep DefectDojo as flexible as possible so you can change DefectDojo vs changing how you do things where you're working. :+1: 

I'm going to go ahead and close this issue but feel free to reopen it if needed.
